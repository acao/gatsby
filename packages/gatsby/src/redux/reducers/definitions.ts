import { ActionsUnion, IGatsbyState } from "../types"

export const definitionsReducer = (
  state: IGatsbyState["definitions"] = new Map(),
  action: ActionsUnion
): IGatsbyState["definitions"] => {
  switch (action.type) {
    case `SET_DEFINITIONS`:
      return action.payload
    default:
      return state
  }
}
