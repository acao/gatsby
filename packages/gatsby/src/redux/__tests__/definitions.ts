import { definitionsReducer } from "../reducers/definitions"
import { print, DocumentNode } from "graphql"
import gql from "graphql-tag"
import { IDefinitionMeta } from "../types"

import { actions } from "../actions"

const parsedExample = gql`
  type DummyType {
    nothing: String
  }
  fragment Example on Query {
    example
  }
`

const parsedExample2 = gql`
  type DummyType {
    nothing: String
  }
  fragment Example2 on Query {
    example2
  }
`

const getMockDefinitions = (
  ast: DocumentNode
): Map<string, IDefinitionMeta> => {
  const definitions = new Map<string, IDefinitionMeta>()

  ast.definitions.map(d => {
    if (!(`name` in d) || d.name === undefined) {
      return
    }
    definitions.set(d?.name.toString(), {
      name: d?.name.toString(),
      filePath: `uri/example`,
      isFragment: Boolean(d.kind === `FragmentDefinition`),
      printedAst: print(d),
      def: d,
      isHook: false,
      isStaticQuery: true,
      hash: `234234`,
      text: ``,
      templateLoc: 0,
    } as IDefinitionMeta)
  })
  return definitions
}

const mockDefinitions = getMockDefinitions(parsedExample)
const mockDefinitions2 = getMockDefinitions(parsedExample2)

describe(`Definitions actions/reducer`, () => {
  it(`allows setting definitionsByName Map`, () => {
    expect(actions.setGraphQLDefinitions(mockDefinitions)).toMatchSnapshot()
  })
  it(`allows updating status`, () => {
    let state = definitionsReducer(
      undefined,
      actions.setGraphQLDefinitions(mockDefinitions)
    )
    state = definitionsReducer(
      state,
      actions.setGraphQLDefinitions(mockDefinitions2)
    )
    expect(state).toEqual(mockDefinitions2)
  })
})
