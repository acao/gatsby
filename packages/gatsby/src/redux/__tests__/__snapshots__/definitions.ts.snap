// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Definitions actions/reducer allows setting definitionsByName Map 1`] = `
Object {
  "payload": Map {
    "[object Object]" => Object {
      "def": Object {
        "directives": Array [],
        "kind": "FragmentDefinition",
        "name": Object {
          "kind": "Name",
          "value": "Example",
        },
        "selectionSet": Object {
          "kind": "SelectionSet",
          "selections": Array [
            Object {
              "alias": undefined,
              "arguments": Array [],
              "directives": Array [],
              "kind": "Field",
              "name": Object {
                "kind": "Name",
                "value": "example",
              },
              "selectionSet": undefined,
            },
          ],
        },
        "typeCondition": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "Query",
          },
        },
      },
      "filePath": "uri/example",
      "hash": "234234",
      "isFragment": true,
      "isHook": false,
      "isStaticQuery": true,
      "name": "[object Object]",
      "printedAst": "fragment Example on Query {
  example
}",
      "templateLoc": 0,
      "text": "",
    },
  },
  "type": "SET_GRAPHQL_DEFINITIONS",
}
`;
